google_api_lib/google_api_client.py

```diff
@@ -0,0 +1,214 @@
+                    'fields': 'userEnteredFormat.textFormat.fontFamily,'
+                              'userEnteredValue.stringValue,'
+                              'userEnteredValue.formulaValue,'
+                              'userEnteredFormat.textFormat.bold',
+                    'range': {
+                        'sheetId': sheet_id,
+                        'startRowIndex': 0,
+                        'endRowIndex': 4 + len(feeders),
+                        'startColumnIndex': 0,
+                        'endColumnIndex': 5,
+                    },
+                    'rows': [
+                        {'values': [{'userEnteredValue': {'stringValue': 'THIS PAGE IS AUTOGENERATED AND WILL BE OVERWRITTEN WHEN THERE ARE PUZZLE UPDATES.'}}]},
+                        {'values': [{'userEnteredValue': {'stringValue': 'DO NOT MAKE CHANGES HERE; COPY INFO TO ANOTHER SHEET FIRST.'}}]},
+                        {},
+                        {'values': [
+                            {'userEnteredValue': {'stringValue': 'Puzzle Link'}, 'userEnteredFormat': {'textFormat': {'bold': True}}},
+                            {'userEnteredValue': {'stringValue': 'Puzzle Name'}, 'userEnteredFormat': {'textFormat': {'bold': True}}},
+                            {'userEnteredValue': {'stringValue': 'Answer'}, 'userEnteredFormat': {'textFormat': {'bold': True}}},
+                            {'userEnteredValue': {'stringValue': 'Sheet'}, 'userEnteredFormat': {'textFormat': {'bold': True}}},
+                            {'userEnteredValue': {'stringValue': 'Slack'}, 'userEnteredFormat': {'textFormat': {'bold': True}}},
+                        ]},
+                    ] + [
+                        {'values': [
+                            {'userEnteredValue': {'formulaValue': '=HYPERLINK("%s", "puzzle")' % puzzle.url}},
+                            {'userEnteredValue': {'stringValue': puzzle.name}},
+                            {'userEnteredValue': {'stringValue': puzzle.answer}, 'userEnteredFormat': {'textFormat': {'fontFamily': 'Roboto Mono'}}},
+                            {'userEnteredValue': {'formulaValue': '=HYPERLINK("%s", "sheet")' % puzzle.sheet}},
+                            {'userEnteredValue': {'formulaValue': '=HYPERLINK("%s/app_redirect?channel=%s", "channel")' % (settings.SLACK_BASE_URL, puzzle.channel)}},
+                        ]} for puzzle in feeders
+                    ],
+                },
+            }, {
+                'autoResizeDimensions': {
+                'dimensions': {
+                    'sheetId': sheet_id,
+                    'dimension': 'COLUMNS',
+                    'startIndex': 1,
+                    'endIndex': 3,
+                },
+            },
+            }]
+        }
+
+        self._sheets_service.spreadsheets().batchUpdate(
+            spreadsheetId=spreadsheet_id,
+            body=body
+        ).execute(http=http)
+
+    @staticmethod
+    def update_meta_sheet_feeders(meta_puzzle):
```

Comment:
One race condition with this:

Suppose 2 people make updates:
- Update 1 starts updating the spreadsheet
- Update 2 starts updating and finishes updating the spreadsheet
- Update 1 commits to the spreadsheet

The spreadsheet ends up in a stale final state until the next update. You could use redlock as a distributed lock to solve this: https://blog.heroku.com/heroku-redis-now-available-in-public-beta

Reply:
Lol, you really like Redis, huh?

You're right that there are race conditions and also we don't retry or rollback anything in case of failures midway. I think the proper way to solve these problems is to put this background work in a queue and use worker processes with some sort of locking (perhaps Redlock, perhaps something else). I'll add this to a TODO as well.

I think it's unlikely we'll encounter many race conditions in practice since on average, and even when they happen, the consequences aren't very dire -- we can always make a dummy change to retrigger updating the meta spreadsheet.

Reply:
Another simple way of solving the race conditions is just to use a thread pool with 1 thread or when we migrate to background processes, a single worker process. 1 thread/process should be sufficient for our purposes.

Reply:
Updated in https://github.com/cardinalitypuzzles/smallboard/pull/140/commits/24b088042c834f906e1ca33f61f2858822977752 to use a ThreadPoolExecutor with 1 worker.

Reply:
Also added TODO to migrate to work queue and separate worker process in future.

Reply:
Ah, ok, this probably works fine for our deployment, but you will need locking with multiple Django instances or multiple worker queues.

Good point on retrying on failure too.

> I think it's unlikely we'll encounter many race conditions in practice since on average, and even when they happen, the consequences aren't very dire -- we can always make a dummy change to retrigger updating the meta spreadsheet.

Yeah, definitely not serious, and solvable with a workaround as you mention.

Reply:
Oh, good point, I didn't think about the case of having multiple dyno instances, which we plan to have during the hunt.

I added a comment about needing locking: https://github.com/cardinalitypuzzles/smallboard/commit/ea27a7e6413291814300a11f7914b8bee0af78cd